### PR TITLE
EL-2897 Moved NavigationModule from ux-aspects-micro-focus to UXAspects.

### PR DESCRIPTION
**Pull request by PearseMcMurray**

### CODE REVIEW

File: src/components/navigation/navigation-item/navigation-item.component.html [hunk 1]
```diff
@@ -0,0 +1,7 @@
+<a *ngIf="link" (click)="toggle(); $event.stopPropagation()" [class.has-arrow]="this.children.length > 0" [class.no-arrow]="this.indentWithoutArrow()" [routerLink]="link">
```
Comment (donna-marijne) [inline src/components/navigation/navigation-item/navigation-item.component.html]: You don't toggle if you have a link - NavigationEnd should handle the expansion of parents.

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+  QueryList,
+  Renderer2,
+  ElementRef,
+  Optional,
+  SkipSelf,
+  AfterContentInit,
+  ContentChild,
+  Output,
+  AfterViewInit,
+  HostListener
```
Comment (donna-marijne) [inline src/components/navigation/navigation-item/navigation-item.component.ts]: Make indentation consistent. We use 4 spaces for typescript files in this repository.

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    encapsulation: ViewEncapsulation.None
+})
+export class NavigationItemComponent implements AfterViewInit {
+    @Input() header: string;
+    @Input() icon: string;
+    @Input() expanded: boolean = false;
+    @Input() link?: string;
+
+    @Input()
+    set level(value: number) {
```
Comment (donna-marijne) [inline src/components/navigation/navigation-item/navigation-item.component.ts]: What does it mean to set `level` as an input property? The level should be calculated from the hierarchy.

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    get level(): number {
+      return this._level;
+    }
+
+    @Output()
+    get active() {
+      if (this.link) {
+          return this._router.isActive(this.link, true);
+      } else {
+          return this._active;
```
Comment (donna-marijne) [inline src/components/navigation/navigation-item/navigation-item.component.ts]: I feel like we had this discussion before - what is the use case for a nav item being active other than being the active route?

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+      }
+    }
+
+    private _level: number = 2;
```
Comment (donna-marijne) [inline src/components/navigation/navigation-item/navigation-item.component.ts]: Why is this 2 by default?

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        private _elementRef: ElementRef,
+        private _renderer: Renderer2,
+        @Optional()
+        @SkipSelf()
+        private _parent: NavigationItemComponent,
+        @Optional() private _router: Router
+    ) {
+        this.level = _parent ? _parent.level + 1 : 2;
+
+        if (this._router) {
```
Comment (donna-marijne) [inline src/components/navigation/navigation-item/navigation-item.component.ts]: I don't understand this. Router is a part of Angular. Why do we need to check for it?

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+      return false;
+    }
+
+    private getLevelClass(): string {
+      var levelClass: string = '';
+      switch (this._level) {
+        case 2:
+          levelClass = 'nav-second-level';
+          break;
```
Comment (donna-marijne) [inline src/components/navigation/navigation-item/navigation-item.component.ts]: Make this more compact by doing e.g. `return 'nav-second-level'`.

File: src/components/navigation/navigation.component.html [hunk 8]
```diff
@@ -0,0 +1,12 @@
+<nav class="navbar-default navbar-static-side-dark tree">
+    <ol class="nav">
+        <ng-content></ng-content>
+    </ol>
+</nav>
+
+<br>
+<br>
+<hr>
+<br>
```
Comment (donna-marijne) [inline src/components/navigation/navigation.component.html]: Remove the above 4 lines.

File: src/components/navigation/navigation-item/navigation-item.component.html (hunk 1 above)
Comment (ashley-hunter) [inline src/components/navigation/navigation-item/navigation-item.component.html]: Also, remove the use of `this` in the view, it is not needed as the component class is the context used for rendering the view

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+  Component,
+  Input,
+  ViewEncapsulation,
+  ContentChildren,
+  QueryList,
+  Renderer2,
+  ElementRef,
+  Optional,
+  SkipSelf,
+  AfterContentInit,
```
Comment (ashley-hunter) [inline src/components/navigation/navigation-item/navigation-item.component.ts]: Review imports, some of these are never used

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+    private hasActiveLink(link: string | UrlTree): boolean {
+      if (link) {
+        return this._router.isActive(link, true) || (this._parent && this._parent.hasActiveLink(link));
+      }
+
+      return false;
+    }
+
+    private getLevelClass(): string {
+      var levelClass: string = '';
```
Comment (ashley-hunter) [inline src/components/navigation/navigation-item/navigation-item.component.ts]: Should never use `var` any more. If the value never changes use `const` otherwise use `let`

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+    }
+
+    ngAfterViewInit() {
+      let ol = this._elementRef.nativeElement.querySelector('.nav');
+      if (ol) {
+        this._renderer.addClass(ol, this.getLevelClass());
+      }
+    }
+
+    toggle() {
```
Comment (ashley-hunter) [inline src/components/navigation/navigation-item/navigation-item.component.ts]: Add return types to functions, both for consistency and also we potentially get better error detection if changes are made in future

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+        '[class.active]': 'active',
+        '[class.selected]': 'expanded',
+    },
+    encapsulation: ViewEncapsulation.None
+})
+export class NavigationItemComponent implements AfterViewInit {
+    @Input() header: string;
+    @Input() icon: string;
+    @Input() expanded: boolean = false;
+    @Input() link?: string;
```
Comment (ashley-hunter) [inline src/components/navigation/navigation-item/navigation-item.component.ts]: Remove the `?`. By default the value will be undefined unless the user specifies it.

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+                this.select();
+            }
+
+            this.expanded = this.hasActiveLink(this.link);
+          });
+        }
+    }
+
+    ngAfterViewInit() {
+      let ol = this._elementRef.nativeElement.querySelector('.nav');
```
Comment (ashley-hunter) [inline src/components/navigation/navigation-item/navigation-item.component.ts]: Use `const` and also possibly rename this variable to give better indication about what we are searching for

**Comment by donna-marijne** [comment]
Pending resolution to https://github.houston.softwaregrp.net/caf/quantum/pull/358.

File: src/components/navigation/navigation-item/navigation-item.component.html (hunk 1 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.html]: toggle() removed from anchor element with router link.

File: src/components/navigation/navigation-item/navigation-item.component.ts (hunk 2 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: Changed indentation to four spaces in all files.

File: src/components/navigation/navigation-item/navigation-item.component.ts (hunk 3 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: Unneeded set function for _level variable removed.

File: src/components/navigation/navigation-item/navigation-item.component.ts (hunk 4 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: get active() modified - only active routes can return true.

File: src/components/navigation/navigation-item/navigation-item.component.ts (hunk 5 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: Default value of _level changed to 1. Along with a change to ngAfterViewInit() this will allow the assigment of classes like nav-second-level, nav-third-level, etc to match their distribution in the ng1 example.

File: src/components/navigation/navigation-item/navigation-item.component.ts (hunk 6 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: Removed check for existence of router. This was a remnant of checks I was making earlier in development concerning which router objects (Router, ActivatedRoute, ...) were available in the documentation example and the Plunker example.

File: src/components/navigation/navigation-item/navigation-item.component.ts (hunk 7 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: Reworked getLevelClass().

File: src/components/navigation/navigation.component.html (hunk 8 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation.component.html]: Unneeded lines removed.

File: src/components/navigation/navigation-item/navigation-item.component.html (hunk 1 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.html]: Removed unnecessary 'this' in template.

File: src/components/navigation/navigation-item/navigation-item.component.ts (hunk 9 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: Removed unused imports.

File: src/components/navigation/navigation-item/navigation-item.component.ts (hunk 12 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: '?' removed.

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+    },
+    encapsulation: ViewEncapsulation.None
+})
+export class NavigationItemComponent implements AfterViewInit {
+    @Input() header: string;
+    @Input() icon: string;
+    @Input() expanded: boolean = false;
+    @Input() link: string;
+
+    @Output()
```
Comment (ashley-hunter) [inline src/components/navigation/navigation-item/navigation-item.component.ts]: Can remove @Output as this is not an eventemitter. Also likely dont need _level any more, just make this:

```typescript
level: number = 1;
```

Remove the @output from active as well. Also active getter never uses _active so this can be removed

File: src/components/navigation/navigation-item/navigation-item.component.ts (hunk 13 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: Renamed variable in ngAfterViewInit() and declared as const.

File: src/components/navigation/navigation-item/navigation-item.component.ts (hunk 11 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: Added return types to functions.

File: src/components/navigation/navigation-item/navigation-item.component.ts (hunk 10 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: Replaced var with const or let.

File: src/components/navigation/navigation.component.ts [hunk 15]
```diff
@@ -0,0 +1,10 @@
+import { Component, Input, ViewEncapsulation } from '@angular/core';
+
+@Component({
+    selector: 'ux-navigation',
+    templateUrl: './navigation.component.html',
+    encapsulation: ViewEncapsulation.None
```
Comment (ashley-hunter) [inline src/components/navigation/navigation.component.ts]: We can remove this line and the associated import. This is only required if we specify a styleUrls property here

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+        const parentListElement = this._elementRef.nativeElement.parentElement;
+        if (parentListElement) {
+            let levelClass: string = this.getLevelClass();
+            if (levelClass.length > 0) {
+                this._renderer.addClass(parentListElement, levelClass);
+            }
+        }
+    }
+
+    toggle(): void {
```
Comment (ashley-hunter) [inline src/components/navigation/navigation-item/navigation-item.component.ts]: Remove this function - this can be done in the view which is more performant, eg:

```
(click)="expanded = !expanded"
```

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+        private _elementRef: ElementRef,
+        private _renderer: Renderer2,
+        @Optional()
+        @SkipSelf()
+        private _parent: NavigationItemComponent,
+        @Optional() private _router: Router
+    ) {
+        this._level = _parent ? _parent.level + 1 : 1;
+
+        _router.events.filter(event => event instanceof NavigationEnd).subscribe(event => {
```
Comment (ashley-hunter) [inline src/components/navigation/navigation-item/navigation-item.component.ts]: You need to store the subscription of this and dispose it when the component is destroyed - look into ngOnDestroy lifecycle hook

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+import 'rxjs/add/operator/filter';
+
+@Component({
+    selector: '[ux-navigation-item]',
+    templateUrl: './navigation-item.component.html',
+    host: {
+        '[class.active]': 'active',
+        '[class.selected]': 'expanded',
+    },
+    encapsulation: ViewEncapsulation.None
```
Comment (ashley-hunter) [inline src/components/navigation/navigation-item/navigation-item.component.ts]: Can remove this and associated import

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+    ElementRef,
+    Optional,
+    SkipSelf,
+    ContentChild,
+    Output,
+    AfterViewInit,
+  HostListener
+} from '@angular/core';
+import { Router, RouterLinkActive, NavigationEnd, RouterState, ActivatedRoute, UrlTree } from '@angular/router';
+import 'rxjs/add/operator/filter';
```
Comment (ashley-hunter) [inline src/components/navigation/navigation-item/navigation-item.component.ts]: Instead of using this rxjs import here, it is not recommended to use lettable operators instead. [Read this for an intro](https://blog.hackages.io/rxjs-5-5-piping-all-the-things-9d469d1b3f44).

This would change the import to something like:

```typescript
import { filter } from 'rxjs/operators/filter';

...

_router.events.pipe(filter(event => event instanceof NavigationEnd)).subscribe...
```

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+    Component,
+    Input,
+    ViewEncapsulation,
+    ContentChildren,
+    QueryList,
+    Renderer2,
+    ElementRef,
+    Optional,
+    SkipSelf,
+    ContentChild,
```
Comment (ashley-hunter) [inline src/components/navigation/navigation-item/navigation-item.component.ts]: Import never used

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+    Output,
+    AfterViewInit,
+  HostListener
```
Comment (ashley-hunter) [inline src/components/navigation/navigation-item/navigation-item.component.ts]: Import never used

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 22]
```diff
@@ -0,0 +1,138 @@
...
+} from '@angular/core';
+import { Router, RouterLinkActive, NavigationEnd, RouterState, ActivatedRoute, UrlTree } from '@angular/router';
```
Comment (ashley-hunter) [inline src/components/navigation/navigation-item/navigation-item.component.ts]: RouterLinkActive, RouterState and ActivatedRoute can be removed as they are never used

File: src/components/navigation/navigation-item/navigation-item.component.html [hunk 23]
```diff
@@ -0,0 +1,7 @@
+<a *ngIf="link" [class.has-arrow]="children.length > 0" [class.no-arrow]="indentWithoutArrow()" [routerLink]="link">
```
Comment (ashley-hunter) [inline src/components/navigation/navigation-item/navigation-item.component.html]: Is it possible to determine in code if it should have no arrow and store the result in a variable instead of calling `indentWithoutArrow()` in the view? This would result in this function being called every time change detection occurs which is quite inefficient.

File: src/components/navigation/navigation.component.html [hunk 24]
```diff
@@ -0,0 +1,7 @@
+<nav class="navbar-default navbar-static-side-dark tree" role="navigation">
```
Comment (donna-marijne) [inline src/components/navigation/navigation.component.html]: Remove these outdated classes, as previously requested. (We do not support a dark theme, apart from anything.)

File: src/components/navigation/navigation-item/navigation-item.component.html (hunk 23 above)
Comment (donna-marijne) [reply src/components/navigation/navigation-item/navigation-item.component.html]: Perhaps an observable based on `this._children.changes`?

Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.html]: Information now stored in variable which is assigned a value in ngAfterContentInit().

File: src/components/navigation/navigation-item/navigation-item.component.ts (hunk 20 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: Import removed.

File: src/components/navigation/navigation-item/navigation-item.component.ts (hunk 21 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: Import removed.

File: src/components/navigation/navigation-item/navigation-item.component.ts (hunk 22 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: Imports removed.

File: src/components/navigation/navigation-item/navigation-item.component.ts (hunk 19 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: Lettable operator used as suggested.

File: src/components/navigation/navigation-item/navigation-item.component.ts (hunk 18 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: encapsulation and import removed.

File: src/components/navigation/navigation-item/navigation-item.component.ts (hunk 14 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: @Output removed from active().
level now a public member. level() removed.

File: src/components/navigation/navigation-item/navigation-item.component.ts (hunk 17 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: Subscription now stored and disposed of in ngOnDestroy().

File: src/components/navigation/navigation-item/navigation-item.component.ts (hunk 16 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: Function removed and view updated as suggested.

File: src/components/navigation/navigation.component.ts (hunk 15 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation.component.ts]: encapsulation and associated import removed.

File: src/components/navigation/navigation.component.html (hunk 24 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation.component.html]: navbar-default and navbar-static-side-dark removed.
Tree class still needed to allow correct styling in MF. Without it, top-level classes will have height 56px instead of the correct 40px.

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 25]
```diff
@@ -61,9 +54,9 @@ export class NavigationItemComponent implements AfterViewInit {
         private _parent: NavigationItemComponent,
         @Optional() private _router: Router
     ) {
-        this._level = _parent ? _parent.level + 1 : 1;
+        this.level = _parent ? _parent.level + 1 : 1;
 
-        _router.events.filter(event => event instanceof NavigationEnd).subscribe(event => {              
+        this._navigationEnd = _router.events.pipe(filter(event => event instanceof NavigationEnd)).subscribe(event => {
```
[the remark this answers is not in the document]
Comment (ashley-hunter) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: There are quite a few benefits outlined in the link below - pretty much the way we were doing before is not tree-shakable. Plus I believe the way we were doing before will be removed in rxjs v6 which will be used by Angular v6 by March/April

https://github.com/ReactiveX/rxjs/blob/master/doc/pipeable-operators.md#why

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 26]
```diff
@@ -73,20 +66,25 @@ export class NavigationItemComponent implements AfterViewInit {
...
+                this._renderer.addClass(parentListElement, 'nav');
                 this._renderer.addClass(parentListElement, levelClass);
             }
         }
     }
 
-    toggle(): void {
-        this.expanded = !this.expanded;
+    ngAfterContentInit(): void {
+        this.setIndentWithoutArrow();
```
[the remark this answers is not in the document]
Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: Subscription added. Sorry, I missed the significance of this change.

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 27]
```diff
@@ -0,0 +1,142 @@
...
+        return this._children.filter(item => item !== this);
+    }    
+
+    constructor(
+        private _elementRef: ElementRef,
+        private _renderer: Renderer2,
+        @Optional()
+        @SkipSelf()
+        private _parent: NavigationItemComponent,
+        @Optional() private _router: Router
```
Comment (donna-marijne) [inline src/components/navigation/navigation-item/navigation-item.component.ts]: Remove Optional here.

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 28]
```diff
@@ -0,0 +1,142 @@
...
+        // Update 'indentWithoutArrow' in response to changes to children
+        this._childrenChanges = this._children.changes.subscribe(() => this.setIndentWithoutArrow());
+    }
+
+    ngOnDestroy () {
+        this._navigationEnd.unsubscribe();
+        this._childrenChanges.unsubscribe();
+    }
+
+    select(): void {
```
Comment (donna-marijne) [inline src/components/navigation/navigation-item/navigation-item.component.ts]: This function is not needed as far as I can tell. Expanding parents should be taken care of by the code above:

````
this.expanded = this.hasActiveLink(this.link);
````

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 29]
```diff
@@ -0,0 +1,142 @@
...
+        private _renderer: Renderer2,
+        @Optional()
+        @SkipSelf()
+        private _parent: NavigationItemComponent,
+        @Optional() private _router: Router
+    ) {
+        this.level = _parent ? _parent.level + 1 : 1;
+
+        this._navigationEnd = _router.events.pipe(filter(event => event instanceof NavigationEnd)).subscribe(event => {              
+            if (this.link && this._router.isActive(this.link, true)) {
```
Comment (donna-marijne) [inline src/components/navigation/navigation-item/navigation-item.component.ts]: Remove this. `this.expanded = this.hasActiveLink(this.link);` takes care of it.

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 30]
```diff
@@ -0,0 +1,142 @@
...
+        '[class.selected]': 'expanded',
+    }
+})
+export class NavigationItemComponent implements AfterViewInit, AfterContentInit, OnDestroy {
+    @Input() header: string;
+    @Input() icon: string;
+    @Input() expanded: boolean = false;
+    @Input() link: string;
+
+    get active() {
```
Comment (donna-marijne) [inline src/components/navigation/navigation-item/navigation-item.component.ts]: Specify return type: boolean.

File: src/components/navigation/navigation-item/navigation-item.component.ts (hunk 27 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: @Optional decorator removed.

File: src/components/navigation/navigation-item/navigation-item.component.ts (hunk 28 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: select() function removed.

File: src/components/navigation/navigation-item/navigation-item.component.ts (hunk 29 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: select() function removed.

File: src/components/navigation/navigation-item/navigation-item.component.ts (hunk 30 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: Return type added to active() getter.

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 31]
```diff
@@ -0,0 +1,135 @@
...
+        @SkipSelf()
+        private _parent: NavigationItemComponent,
+        private _router: Router
+    ) {
+        this.level = _parent ? _parent.level + 1 : 1;
+
+        this._navigationEnd = _router.events.pipe(filter(event => event instanceof NavigationEnd)).subscribe(event => {              
+            this.expanded = this.hasActiveLink(this.link);
+            if (this.expanded) {
+                // Expand ancestors
```
Comment (donna-marijne) [inline src/components/navigation/navigation-item/navigation-item.component.ts]: I don't think this is needed. All navigation items subscribe to this event, and expand as required, so this would be duplicated effort.

File: src/components/navigation/navigation.component.html [hunk 32]
```diff
@@ -0,0 +1,7 @@
+<nav class="tree" role="navigation">
+    <ol class="nav">
+        <ng-content></ng-content>
+    </ol>
+</nav>
+
+<router-outlet></router-outlet>
```
Comment (donna-marijne) [inline src/components/navigation/navigation.component.html]: Isn't it up to the application to determine where the router outlet goes? Why was this added here?

**Comment by donna-marijne** [comment]
Any update on this PR coming soon?

**Comment by PearseMcMurray** [comment]
I haven't had much time to work on this today so I won't be able to submit this today. I have been having problems with the expansion of parent folders when lowest-level items are selected. Can this be moved to the next sprint? After today's planning session I have free time.

**Comment by donna-marijne** [comment]
Done

**Comment by PearseMcMurray** [comment]
Moved router-outlet directive to documentation.
hasActiveLink() now sets the 'expanded' property. Then calls hasActiveLink() in ancestors to set the 'expanded' property in them also. Previously it was not being set in the ancestors.

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 33]
```diff
@@ -0,0 +1,133 @@
...
+
+    ngOnDestroy () {
+        this._navigationEnd.unsubscribe();
+        this._childrenChanges.unsubscribe();
+    }
+
+    private hasActiveLink(link: string | UrlTree): void {
+        if (link) {
+            this.expanded = this._router.isActive(link, true);            
+            if (this.expanded && this._parent) {
```
Comment (donna-marijne) [inline src/components/navigation/navigation-item/navigation-item.component.ts]: This is still wrong. Revert to the original version and please address the review comment. Contact me for clarification if you do not understand.

Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: I have reverted to the original code but expansion of parents of a selected item does not occur, whether they themselves have a link or not. I have experimented with addition and removal of router links in the example HTML and with changes to the link definitions in the router module.

The call to hasActiveLink in the NavigationEnd handler sets the value of the 'expanded' flag according to the component's link. For parents of selected items this is always false. The recursive call within hasActiveLink will return 'true' but it doesn't set the expanded flag in the component.

What am I missing here?

Comment (donna-marijne) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: Something like this?

````
private hasActiveLink(link: string | UrlTree): boolean {
    if (link) {
        return this._router.isActive(link, true) || this.children.any((item) => item.hasActiveLink(item.link));
    }

    return false;
}
````

Explanation: Every component receives the NavigationEnd event. We use this to determine the state of the `expanded` property for each node, ensuring that the entire tree is always up to date with every navigation event. This avoids the need to go back and collapse siblings explicitly. This makes use of recursion, which is okay since the maximum depth is 5, and keeps the code tidy.

Let me know if there is anything I've overlooked with this, or anything that you don't understand. It's not a problem to ask for help, as previously agreed.

**Comment by PearseMcMurray** [comment]
'expanded' property set if component is the active link or if any of its descendants are the active link.

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 34]
```diff
@@ -90,17 +90,22 @@ export class NavigationItemComponent implements AfterViewInit, AfterContentInit,
...
 
-            return;
+        // If this component has children, check if any of them, or their descendants, are active.
+        let hasActiveChildren = false;
+        if (this.children.length > 0) {
+            hasActiveChildren = this.children.some((item) => item.hasActiveLink(item.link));
         }
 
-        this.expanded = false;
+        return hasActiveChildren;
```
Comment (donna-marijne) [inline src/components/navigation/navigation-item/navigation-item.component.ts]: It's fine if you want to write this out for your own understanding, but try to condense it and remove redundant variables, since the browser ultimately has to download the javascript version of this.

This would be more readable and concise:

````
if (link && this._router.isActive(link, true)) {
  return true;
}

return this.children.some((item) => item.hasActiveLink(item.link));
````

**Comment by donna-marijne** [comment]
@ashley-hunter Could you verify these changes and merge when all comments have been addressed to the appropriate standard? Let's try get this done today.

**Comment by ashley-hunter** [comment]
@donna-marijne sure! No problem!

File: src/components/navigation/navigation-item/navigation-item.component.ts (hunk 34 above)
Comment (PearseMcMurray) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: Condensed code as suggested.

File: src/components/navigation/navigation-item/navigation-item.component.ts [hunk 35]
```diff
@@ -0,0 +1,130 @@
...
+    AfterViewInit,
+    AfterContentInit,
+    OnDestroy
+} from '@angular/core';
+import { Router, NavigationEnd, UrlTree } from '@angular/router';
+import { filter } from 'rxjs/operators/filter';
+import { Subscription } from 'rxjs/Subscription';
+
+@Component({
+    selector: '[ux-navigation-item]',
```
Comment (ashley-hunter) [inline src/components/navigation/navigation-item/navigation-item.component.ts]: Attribute selectors should follow upper camel case naming convention, eg: `uxNavigationItem`

Comment (ashley-hunter) [reply src/components/navigation/navigation-item/navigation-item.component.ts]: Leave this as is for now, this is a bit of a grey area as components generally don't use an attribute selector. We can change this at a later point if we want to
